### PR TITLE
[WIP] Ring closure

### PR DESCRIPTION
**Pull request by dominicrufa**
Adding functions to allow for ring closures in relative simulations

### CODE REVIEW

File: perses/annihilation/relative.py [hunk 1]
```diff
@@ -66,6 +66,8 @@ def __init__(self,
                  soften_only_new=False,
                  neglected_new_angle_terms = [],
                  neglected_old_angle_terms = [],
+                 special_new_terms = None,
```
Comment (jchodera) [inline perses/annihilation/relative.py]: Should we use a term other than `special`? Do these represent biasing terms that are applied only during proposal, but are corrected for during the RJMC acceptance step?

Comment (dominicrufa) [reply perses/annihilation/relative.py]: special terms is a dict of omitted bonds, angles, and torsions (which are annealed in through AIS), as well as extra angles and torsions that bias  RJ proposals (which are annealed off during the AIS protocol);

File: perses/annihilation/relative.py [hunk 2]
```diff
@@ -104,6 +106,10 @@ def __init__(self,
         neglected_old_angle_terms : list
             list of indices from the HarmonicAngleForce of the old_system for which the geometry engine neglected.
             Hence, these angles must be alchemically deleted for the unique old atoms (reverse lambda protocol)
+        special_new_terms : dict, default None
+            dict of new valence/1,4 terms used for stereochemical/ring strucure stability
```
Comment (jchodera) [inline perses/annihilation/relative.py]: `strucure` -> `structure`

Can we also add more documentation about how these terms are used? As you note above:
> special terms is a dict of omitted bonds, angles, and torsions (which are annealed in through AIS), as well as extra angles and torsions that bias RJ proposals (which are annealed off during the AIS protocol);

File: perses/annihilation/relative.py [hunk 3]
```diff
@@ -104,6 +106,10 @@ def __init__(self,
...
+        special_old_terms : dict, default None
```
Comment (jchodera) [inline perses/annihilation/relative.py]: Same here as above.

File: perses/annihilation/relative.py (hunk 2 above)
Comment (dominicrufa) [reply perses/annihilation/relative.py]: will do!

File: perses/annihilation/relative.py (hunk 3 above)
Comment (dominicrufa) [reply perses/annihilation/relative.py]: https://github.com/choderalab/perses/pull/585/commits/518105bb9e33e8de9403596e2d21d02322cc78a0

**Comment by hannahbrucemacdonald** [comment]
If I merge conflicts can this be merged?

**Comment by dominicrufa** [comment]
> If I fix conflicts can this be merged?

let me give it a read tonight and write tests (in case i forgot any); i realized i have to change a few things in the `factory`

**Comment by jchodera** [comment]
Happy to chat about algorithms if needed!

**Comment by dominicrufa** [comment]
> Happy to chat about algorithms if needed!

i'm concerned about the following: if we are adding a new ring (i.e. all atoms in the ring are new, would it be advisable to simply not anneal in the final bond that would close the ring over the protocol, and instead just place a lambda-independent bond there?)

my other concern is that all ring closures happening with the PR now add biasing torsions that attempt to make the ring planar (regardless of whether it is aromatic), which is not generally what we want.  however, this would involve passing an oemol to the geometry engine to check whether the ring is aromatic or not (this is also what has to happen to check the chirality of certains atoms for biasing torsions); if the ring is not aromatic, then do i simply not add biasing torsions?

**Comment by hannahbrucemacdonald** [comment]
Based on some misunderstandings that I've had, I think it is important to have clarity about what 'ring closure' means -- does this constitute replacing X-membered ring (all unique_old) with Y-membered ring (all unique new) i.e. entire ring for entire ring substitution OR turning (x-ethyl,(x+1)-ethyl blah into benzene) i.e.e ring with non-ring substitutioin which I think are two distinct issues, with their own issues

**Pull request closed without merging**